_includes/markdown/Typescript.md

```diff
@@ -0,0 +1,188 @@
+
+While we'd like none of these escape hatches to be used, it would be unrealistic to expect all engineers to have types 100% correct on every project all the time. We believe that these recommendations provide a path to address them organically during the life-cycle of the project.
+
+## Tools
+
+- [TypeScript Error Translator](https://ts-error-translator.vercel.app/): Lets you drop any TypeScript error message and gives you a human-readable explanation in plain English.
+- [pretty-ts-errors] (https://marketplace.visualstudio.com/items?itemName=yoavbls.pretty-ts-errors): A VSCode extension that translaters TypeScript errors in a human-readable way right into VSCode.
+
+## TypeScript and CI
+
+We recommend that all projects run the typescript compiler to check type safety at the CI level. Every Pull Request/Merge Request should run `tsc --noEmit` on CI to ensure there are no compilation issues before merging the branch.
+
+This is a great strategy to avoid builds failing due to type errors. With Next.js for instance, you might not realize you have a TS compilation issue until a build is executed, therefore to avoid merging code that produces failing builds we recommend always ensuring that the code has no TypeScript compilation issues before merging.
+
+## Recommended `tsconfig.json` settings
+
+This is not a full list but we highlight some recommended settings here.
+
+1. `noFallthroughCasesInSwitch: true`- [Definition](https://www.typescriptlang.org/tsconfig/#noFallthroughCasesInSwitch) - ensures you are not skipping anything in a switch statement, and always at least provide a default option.
+2. `noImplicitAny: true` - [Definition](https://www.typescriptlang.org/tsconfig/#noImplicitAny) - ensures you define `any` explicitly, rather than implicitly assume it. This makes it a lot easier to find all code relying on any for future refactoring.
+3. `strictNullChecks: true` - [Definition](https://www.typescriptlang.org/tsconfig/#strictNullChecks) - forces handling both `undefined` and `null`
+4. `noUnusedParameters: true` - [Definition](https://www.typescriptlang.org/tsconfig/#noUnusedParameters) - does not allow unused parameters to remain in code
+
+We also have a recommended [tsconfig.json](https://github.com/10up/headstartwp/blob/develop/projects/wp-nextjs/tsconfig.json) for Next.js projects.
+
+## TypeScript and React
+
+### Use `type` for React Props
+
+While `interface` and `type` are mostly interchangeable, prefer `type` over `interfaces` unless you expect the props to be the base for other components's props.
+
+```ts
+type BlocksProps = {
+	html: string;
+};
+
+export const Blocks = ({ html }: BlocksProps) => {
+    // render blocks
+}
+```
+
+### Use function default arguments instead of `defaultProps`
+
+`defaultProps` has been deprecated in React 19, therefore prefer using function default arguments.
+
+```ts
+type MyComponentProps = {
+    title?: string;
```

Comment:
@nicholasio the optional operator here has always bugged me - the semantics makes it difficult to really determine what props are actually required for a component to function in a fundamental way. Is there guidance around this? When should a type leverage an optional prop vs an optional prop with a default? The two are not exclusive and can lead to overuse which makes components brittle

Reply:
I do not think this is a problem that is limited to TS. For example, you still have to decide the same thing when writing OOP PHP. 

In my opinion all keys of an interface should be required and no functions should have default values until you have a good reason to do either. Some examples I can think of
- If you have a React component that accepts a description field but in some cases you don't want to show a description that would be a good example of using an optional key in an interface. Then make sure the code has {description ? <p>{description}</> : null}
- If you have a Button component that accepts the type (submit, button, reset) then probably a good argument for type to have a default 

I agree with you that overuse can create problems which is why I think you need a good reason to use them. And it's important if you use them that you have some sort of way automating testing of the function or component when you supply all the keys and only the required keys (e.g. vrt if it's storybook, jest if it's functions, etc)

Reply:
I think that would out of scope for this particular PR.